File: apps/macos_native/Sources/PasswordManagerMacOSApp/Models/FieldReferenceResolution.swift
import Foundation

enum FieldReferenceStatus: Equatable, Sendable {
    case empty
    case invalidConfiguration
    case missing
    case deleted
    case categoryMismatch
    case targetFieldMissing
    case targetFieldUnsupported
    case targetFieldEmpty
    case resolved
}

struct FieldReferenceTarget: Equatable, Sendable {
    let entryID: String
    let entryLabel: String
    let entryCategory: String
    let fieldID: String
    let fieldName: String
    let fieldValue: String
}

struct FieldReferenceResolution: Equatable, Sendable {
    let status: FieldReferenceStatus
    let target: FieldReferenceTarget?

    init(status: FieldReferenceStatus, target: FieldReferenceTarget? = nil) {
        self.status = status
        self.target = target
    }
}

func resolveFieldReference(
    sourceEntry: VaultEntry,
    field: CustomField,
    categoryTemplates: [CategoryTemplate],
    entries: [VaultEntry]
) -> FieldReferenceResolution? {
    guard let sourceTemplate = matchingFieldReferenceCategoryTemplate(
        category: sourceEntry.payload.category,
        templates: categoryTemplates
    ),
    let sourceTemplateField = customFieldSemantics(
        field: field,
        template: sourceTemplate
    ).templateField,
    sourceTemplateField.valueType == "fieldReference" else {
        return nil
    }

    guard !field.value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
        return FieldReferenceResolution(status: .empty)
    }

    let targetCategory = sourceTemplateField.targetCategory
        .trimmingCharacters(in: .whitespacesAndNewlines)
    let targetFieldID = sourceTemplateField.targetFieldId
    guard !targetCategory.isEmpty,
          !targetFieldID.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
        return FieldReferenceResolution(status: .invalidConfiguration)
    }

    let sourceCategory = sourceTemplate.category.trimmingCharacters(in: .whitespacesAndNewlines)
    guard sourceCategory.caseInsensitiveCompare(targetCategory) != .orderedSame
            || sourceTemplateField.id != targetFieldID else {
        return FieldReferenceResolution(status: .invalidConfiguration)
    }

    guard let targetEntry = entries.first(where: { $0.id == field.value }) else {
        return FieldReferenceResolution(status: .missing)
    }
    let targetEntryCategory = targetEntry.payload.category
        .trimmingCharacters(in: .whitespacesAndNewlines)
    let entryProjection = FieldReferenceTarget(
        entryID: targetEntry.id,
        entryLabel: targetEntry.label,
        entryCategory: targetEntryCategory,
        fieldID: targetFieldID,
        fieldName: "",
        fieldValue: ""
    )
    guard !targetEntry.isDeleted else {
        return FieldReferenceResolution(status: .deleted, target: entryProjection)
    }

    guard targetEntryCategory.caseInsensitiveCompare(targetCategory) == .orderedSame else {
        return FieldReferenceResolution(status: .categoryMismatch, target: entryProjection)
    }

    guard let targetTemplate = matchingFieldReferenceCategoryTemplate(
        category: targetEntryCategory,
        templates: categoryTemplates
    ),
    let targetTemplateField = targetTemplate.fields.first(where: { $0.id == targetFieldID }) else {
        return FieldReferenceResolution(status: .targetFieldMissing, target: entryProjection)
    }
    let fieldProjection = FieldReferenceTarget(
        entryID: targetEntry.id,
        entryLabel: targetEntry.label,
        entryCategory: targetEntryCategory,
        fieldID: targetTemplateField.id,
        fieldName: targetTemplateField.name,
        fieldValue: ""
    )
    guard targetTemplateField.normalizedValueType == "text" else {
        return FieldReferenceResolution(status: .targetFieldUnsupported, target: fieldProjection)
    }

    guard let targetValueField = matchingFieldReferenceTargetValue(
        in: targetEntry.customFields,
        templateField: targetTemplateField
    ),
    !targetValueField.value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
        return FieldReferenceResolution(status: .targetFieldEmpty, target: fieldProjection)
    }

    return FieldReferenceResolution(
        status: .resolved,
        target: FieldReferenceTarget(
            entryID: targetEntry.id,
            entryLabel: targetEntry.label,
            entryCategory: targetEntryCategory,
            fieldID: targetTemplateField.id,
            fieldName: targetTemplateField.name,
            fieldValue: targetValueField.value
        )
    )
}

func propagateFieldReferenceCategoryRename(
    templates: [CategoryTemplate],
    from oldCategory: String,
    to newCategory: String
) -> [CategoryTemplate] {
    let oldNormalized = oldCategory.trimmingCharacters(in: .whitespacesAndNewlines)
    let newNormalized = newCategory.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !oldNormalized.isEmpty, !newNormalized.isEmpty else { return templates }

    return templates.map { template in
        var updatedTemplate = template
        updatedTemplate.fields = template.fields.map { field in
            guard field.valueType == "fieldReference",
                  field.targetCategory.trimmingCharacters(in: .whitespacesAndNewlines)
                    .caseInsensitiveCompare(oldNormalized) == .orderedSame else {
                return field
            }
            var updatedField = field
            updatedField.targetCategory = newNormalized
            return updatedField
        }
        return updatedTemplate
    }
}

func fieldReferenceTargetFieldIDs(
    targetCategory: String,
    templates: [CategoryTemplate]
) -> Set<String> {
    let normalizedCategory = targetCategory.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !normalizedCategory.isEmpty else { return [] }

    return Set(templates.flatMap(\.fields).compactMap { field in
        guard field.valueType == "fieldReference",
              field.targetCategory.trimmingCharacters(in: .whitespacesAndNewlines)
                .caseInsensitiveCompare(normalizedCategory) == .orderedSame,
              !field.targetFieldId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return nil
        }
        return field.targetFieldId
    })
}

extension VaultEntry {
    func withFieldReferenceSearchProjection(
        categoryTemplates: [CategoryTemplate],
        entries: [VaultEntry]
    ) -> VaultEntry {
        let sourceTemplate = matchingFieldReferenceCategoryTemplate(
            category: payload.category,
            templates: categoryTemplates
        )
        var projectedEntry = withEntryReferenceSearchProjection(
            template: sourceTemplate,
            entries: entries
        )
        projectedEntry.customFields = zip(customFields, projectedEntry.customFields).map {
            sourceField, projectedField in
            guard customFieldSemantics(field: sourceField, template: sourceTemplate)
                .templateField?.valueType == "fieldReference" else {
                return projectedField
            }

            var fieldProjection = projectedField
            let resolution = resolveFieldReference(
                sourceEntry: self,
                field: sourceField,
                categoryTemplates: categoryTemplates,
                entries: entries
            )
            if resolution?.status == .resolved, let target = resolution?.target {
                fieldProjection.value = [
                    target.entryLabel,
                    target.entryCategory,
                    target.fieldName
                ]
                .filter { !$0.isEmpty }
                .joined(separator: " ")
            } else {
                fieldProjection.value = ""
            }
            return fieldProjection
        }
        return projectedEntry
    }

    func remappingFieldReferenceIDs(
        using destinationIDsBySourceID: [String: String],
        template: CategoryTemplate?
    ) -> VaultEntry {
        var remappedEntry = self
        remappedEntry.customFields = customFields.map { field in
            guard customFieldSemantics(field: field, template: template)
                .templateField?.valueType == "fieldReference",
            let destinationID = destinationIDsBySourceID[field.value] else {
                return field
            }
            var remappedField = field
            remappedField.value = destinationID
            return remappedField
        }
        return remappedEntry
    }
}

private func matchingFieldReferenceCategoryTemplate(
    category: String,
    templates: [CategoryTemplate]
) -> CategoryTemplate? {
    let normalizedCategory = category.trimmingCharacters(in: .whitespacesAndNewlines)
    return templates.first {
        $0.category.trimmingCharacters(in: .whitespacesAndNewlines)
            .caseInsensitiveCompare(normalizedCategory) == .orderedSame
    }
}

private func matchingFieldReferenceTargetValue(
    in fields: [CustomField],
    templateField: FieldTemplate
) -> CustomField? {
    if let exact = fields.first(where: {
        !$0.templateFieldId.isEmpty && $0.templateFieldId == templateField.id
    }) {
        return exact
    }

    let targetName = templateField.name.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !targetName.isEmpty else { return nil }
    return fields.first {
        $0.templateFieldId.isEmpty
            && $0.name.trimmingCharacters(in: .whitespacesAndNewlines)
                .caseInsensitiveCompare(targetName) == .orderedSame
    }
}

File: apps/android_native/app/src/main/java/life/devops/passwordmanager/model/FieldReferenceResolution.kt
package life.devops.passwordmanager.model

internal enum class FieldReferenceStatus {
    EMPTY,
    INVALID_CONFIGURATION,
    MISSING,
    DELETED,
    CATEGORY_MISMATCH,
    TARGET_FIELD_MISSING,
    TARGET_FIELD_UNSUPPORTED,
    TARGET_FIELD_EMPTY,
    RESOLVED,
}

internal data class FieldReferenceTargetEntry(
    val id: String,
    val label: String,
    val category: String,
)

internal data class FieldReferenceTargetField(
    val id: String,
    val name: String,
    val value: String,
)

internal data class FieldReferenceResolution(
    val status: FieldReferenceStatus,
    val targetEntry: FieldReferenceTargetEntry? = null,
    val targetField: FieldReferenceTargetField? = null,
)

internal fun resolveFieldReference(
    field: CustomField,
    sourceTemplate: CategoryTemplate?,
    categoryTemplates: Collection<CategoryTemplate>,
    entries: Collection<VaultEntry>,
): FieldReferenceResolution? =
    resolveFieldReference(
        field = field,
        sourceTemplate = sourceTemplate,
        findTargetById = { targetId -> entries.firstOrNull { entry -> entry.id == targetId } },
        findTemplateByCategory = { targetCategory ->
            categoryTemplates.firstOrNull { template ->
                template.category.trim().equals(targetCategory, ignoreCase = true)
            }
        },
    )

internal fun resolveFieldReference(
    field: CustomField,
    sourceTemplate: CategoryTemplate?,
    categoryTemplatesByName: Map<String, CategoryTemplate>,
    entriesById: Map<String, VaultEntry>,
): FieldReferenceResolution? =
    resolveFieldReference(
        field = field,
        sourceTemplate = sourceTemplate,
        findTargetById = entriesById::get,
        findTemplateByCategory = { category -> categoryTemplatesByName[category.lowercase()] },
    )

private inline fun resolveFieldReference(
    field: CustomField,
    sourceTemplate: CategoryTemplate?,
    findTargetById: (String) -> VaultEntry?,
    findTemplateByCategory: (String) -> CategoryTemplate?,
): FieldReferenceResolution? {
    val sourceTemplateField = fieldReferenceTemplateField(field, sourceTemplate) ?: return null
    if (field.value.isBlank()) {
        return FieldReferenceResolution(status = FieldReferenceStatus.EMPTY)
    }

    val targetCategory = sourceTemplateField.targetCategory.trim()
    val targetFieldId = sourceTemplateField.targetFieldId
    if (
        targetCategory.isEmpty() ||
        targetFieldId.isBlank() ||
        (
            sourceTemplate?.category.orEmpty().trim().equals(targetCategory, ignoreCase = true) &&
                sourceTemplateField.id == targetFieldId
        )
    ) {
        return FieldReferenceResolution(status = FieldReferenceStatus.INVALID_CONFIGURATION)
    }

    val target = findTargetById(field.value)
        ?: return FieldReferenceResolution(status = FieldReferenceStatus.MISSING)
    val targetEntry = FieldReferenceTargetEntry(
        id = target.id,
        label = target.label,
        category = target.payload.category.trim(),
    )
    if (target.isDeleted) {
        return FieldReferenceResolution(
            status = FieldReferenceStatus.DELETED,
            targetEntry = targetEntry,
        )
    }
    if (!targetEntry.category.equals(targetCategory, ignoreCase = true)) {
        return FieldReferenceResolution(
            status = FieldReferenceStatus.CATEGORY_MISMATCH,
            targetEntry = targetEntry,
        )
    }

    val targetTemplate = findTemplateByCategory(targetCategory)
        ?: return FieldReferenceResolution(
            status = FieldReferenceStatus.TARGET_FIELD_MISSING,
            targetEntry = targetEntry,
        )
    val targetTemplateField = targetTemplate.fields.firstOrNull { templateField ->
        templateField.id == targetFieldId
    } ?: return FieldReferenceResolution(
        status = FieldReferenceStatus.TARGET_FIELD_MISSING,
        targetEntry = targetEntry,
    )
    val targetField = FieldReferenceTargetField(
        id = targetTemplateField.id,
        name = targetTemplateField.name.trim(),
        value = "",
    )
    if (targetTemplateField.normalizedValueType() != CUSTOM_FIELD_TEXT_VALUE_TYPE) {
        return FieldReferenceResolution(
            status = FieldReferenceStatus.TARGET_FIELD_UNSUPPORTED,
            targetEntry = targetEntry,
            targetField = targetField,
        )
    }
    if (targetTemplateField.isBuiltInEntryNameField()) {
        if (target.label.isBlank()) {
            return FieldReferenceResolution(
                status = FieldReferenceStatus.TARGET_FIELD_EMPTY,
                targetEntry = targetEntry,
                targetField = targetField,
            )
        }
        return FieldReferenceResolution(
            status = FieldReferenceStatus.RESOLVED,
            targetEntry = targetEntry,
            targetField = targetField.copy(value = target.label),
        )
    }

    val targetValue = matchingTargetFieldValue(target, targetTemplateField)
    if (targetValue == null || targetValue.value.isBlank()) {
        return FieldReferenceResolution(
            status = FieldReferenceStatus.TARGET_FIELD_EMPTY,
            targetEntry = targetEntry,
            targetField = targetField,
        )
    }
    return FieldReferenceResolution(
        status = FieldReferenceStatus.RESOLVED,
        targetEntry = targetEntry,
        targetField = targetField.copy(value = targetValue.value),
    )
}

internal fun isFieldReference(
    field: CustomField,
    sourceTemplate: CategoryTemplate?,
): Boolean = fieldReferenceTemplateField(field, sourceTemplate) != null

private fun matchingTargetFieldValue(
    target: VaultEntry,
    targetTemplateField: FieldTemplate,
): CustomField? =
    target.customFields.firstOrNull { customField ->
        customField.templateFieldId.isNotEmpty() &&
            customField.templateFieldId == targetTemplateField.id
    } ?: target.customFields.firstOrNull { customField ->
        customField.templateFieldId.isEmpty() &&
            customField.name.trim().isNotEmpty() &&
            customField.name.trim().equals(targetTemplateField.name.trim(), ignoreCase = true)
    }

private fun FieldTemplate.isBuiltInEntryNameField(): Boolean =
    id == CategoryTemplate.stableFieldId("名称") || name.trim() == "名称"

private fun fieldReferenceTemplateField(
    field: CustomField,
    sourceTemplate: CategoryTemplate?,
): FieldTemplate? {
    val templateField = matchingCustomFieldTemplate(field, sourceTemplate) ?: return null
    return templateField.takeIf { it.valueType == FIELD_REFERENCE_VALUE_TYPE }
}

internal const val FIELD_REFERENCE_VALUE_TYPE = "fieldReference"
